File: ios/Classes/UniSdkPlugin.m
//
//  UniSdkPlugin.m
//  uni_sdk_plugin
//
//  Created by Eric on 2023/6/6.
//

#import "UniSdkPlugin.h"
#import "DCUniMP.h"
#import "DCUniMPSDKEngine.h"

@interface UniSdkPlugin ()<DCUniMPSDKEngineDelegate>
@property (nonatomic, strong) FlutterMethodChannel *channel;
@end

@implementation UniSdkPlugin

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
    FlutterMethodChannel *channel = [FlutterMethodChannel methodChannelWithName:@"uni_sdk_plugin" binaryMessenger:[registrar messenger]];
    UniSdkPlugin *instance = [[UniSdkPlugin alloc] init];
    instance.channel = channel;
    [registrar addApplicationDelegate:instance];
    [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
    NSLog(@"%@", call.method);
    if ([call.method isEqualToString:@"openUniMP"]) {
        NSDictionary *arguments = call.arguments;
        [self beforeOpenUniMPWithAppId:arguments[@"appId"] url:arguments[@"url"] version:arguments[@"version"] result:result];
    } else {
        result(FlutterMethodNotImplemented);
    }
}

- (void)onUniMPEventReceive:(NSString *)appid event:(NSString *)event data:(id)data callback:(DCUniMPKeepAliveCallback)callback {
    if ([event isEqualToString:@"getUserInfo"]) {
        [self getUserInfoWithCallback:^(NSDictionary *userInfo) {
            if (userInfo) {
                NSString *uid = userInfo[@"uid"];
                NSString *token = userInfo[@"token"];
                callback(@{@"uid": uid, @"token": token}, NO);
            } else {
                callback(@{}, NO);
            }
        }];
    }
}

- (void)beforeOpenUniMPWithAppId:(NSString *)appid url:(NSString *)url version:(NSString *)version result:(FlutterResult)result {
    NSString *vnName = @"";
    NSDictionary *versionInfo = [DCUniMPSDKEngine getUniMPVersionInfoWithAppid:appid];
    if (versionInfo) {
        vnName = versionInfo[@"name"];
    }
    if ([DCUniMPSDKEngine isExistsUniMP:appid] && [vnName isEqualToString:version]) {
        [self openUniMPWithAppId:appid result:result];
    } else {
        NSString *filename = [[NSURL URLWithString:url] lastPathComponent];
        if (![self isFileExistsInDocuments:filename]) {
            __weak typeof(self) weakSelf = self;
            [self downloadFileAndSaveToDocumentsWithURL:[NSURL URLWithString:url] result:result completion:^{
                [weakSelf installAndOpenUniMPWithAppId:appid url:url result:result];
            }];
        }else{
            [self installAndOpenUniMPWithAppId:appid url:url result:result];
        }
    }
}

- (void)openUniMPWithAppId:(NSString *)appid result:(FlutterResult)result {
    DCUniMPConfiguration *config = [self getUniMPConfiguration];
    [DCUniMPSDKEngine openUniMP:appid configuration:config completed:^(DCUniMPInstance * _Nullable uniMPInstance, NSError * _Nullable error) {
        if (uniMPInstance) {
            NSLog(@"小程序打开成功");
            result(@(YES));
        } else {
            NSLog(@"%@", error);
            result(@(NO));
        }
    }];
}

- (void)installAndOpenUniMPWithAppId:(NSString *)appid url:(NSString *)url result:(FlutterResult)result {
    NSLog(@"install uni mp");
    NSString *docsDir = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES)[0];
    NSString *filePath = [docsDir stringByAppendingPathComponent:[NSURL URLWithString:url].lastPathComponent];
    NSLog(@"%@", filePath);
    
    
    __weak typeof(self) weakSelf = self;
    NSError *installError = nil;
    [DCUniMPSDKEngine installUniMPResourceWithAppid:appid resourceFilePath:filePath password:nil error:&installError];
    if (!installError) {
        [weakSelf openUniMPWithAppId:appid result:result];
    } else {
        NSLog(@"load unimp error");
    }
}

- (DCUniMPConfiguration *)getUniMPConfiguration {
    DCUniMPConfiguration *config = [[DCUniMPConfiguration alloc] init];
    config.extraData = @{@"arguments": @"hello"};
    config.enableBackground = YES;
    return config;
}

//判断文件是否存在
- (BOOL)isFileExistsInDocuments:(NSString *)filename {
    NSFileManager *fileManager = [NSFileManager defaultManager];
    NSURL *documentsURL = [[fileManager URLsForDirectory:NSDocumentDirectory inDomains:NSUserDomainMask] firstObject];
    NSURL *fileURL = [documentsURL URLByAppendingPathComponent:filename];
    
    return [fileManager fileExistsAtPath:fileURL.path];
}

- (void)downloadFileAndSaveToDocumentsWithURL:(NSURL *)url result:(FlutterResult)result completion:(dispatch_block_t)completion {
    NSURLSession *session = [NSURLSession sharedSession];
    NSURLSessionDownloadTask *task = [session downloadTaskWithURL:url completionHandler:^(NSURL *temporaryURL, NSURLResponse *response, NSError *error) {
        if (error) {
            NSLog(@"Error downloading file: %@", error);
            result(@(NO));
            return;
        }
        
        if (temporaryURL) {
            NSFileManager *fileManager = [NSFileManager defaultManager];
            NSURL *documentsURL = [fileManager URLsForDirectory:NSDocumentDirectory inDomains:NSUserDomainMask][0];
            NSURL *destinationURL = [documentsURL URLByAppendingPathComponent:url.lastPathComponent];
            
            NSError *moveError = nil;
            [fileManager moveItemAtURL:temporaryURL toURL:destinationURL error:&moveError];
            if (!moveError) {
                NSLog(@"File downloaded and saved to: %@", destinationURL.path);
                dispatch_async(dispatch_get_main_queue(), ^{
                    completion();
                });
            } else {
                result(@(NO));
                NSLog(@"Error saving file: %@", moveError);
            }
        } else {
            result(@(NO));
            NSLog(@"Temporary URL is nil.");
        }
    }];
    
    [task resume];
}

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
    [DCUniMPSDKEngine initSDKEnvironmentWithLaunchOptions:launchOptions];
    [DCUniMPSDKEngine setDelegate:self];
    return YES;
}


- (void)applicationDidBecomeActive:(UIApplication *)application {
    [DCUniMPSDKEngine applicationDidBecomeActive:application];
}

- (void)applicationWillResignActive:(UIApplication *)application {
    [DCUniMPSDKEngine applicationWillResignActive:application];
}

- (void)applicationDidEnterBackground:(UIApplication *)application {
    [DCUniMPSDKEngine applicationDidEnterBackground:application];
}

- (void)applicationWillEnterForeground:(UIApplication *)application {
    [DCUniMPSDKEngine applicationWillEnterForeground:application];
}

- (void)applicationWillTerminate:(UIApplication *)application {
    [DCUniMPSDKEngine destory];
}

- (void)getUserInfoWithCallback:(void (^)(NSDictionary *))callback {
    [self.channel invokeMethod:@"getUserInfo" arguments:nil result:^(id _Nullable result) {
        NSDictionary *userInfo = (NSDictionary *)result;
        callback(userInfo);
    }];
}

@end

File: ios/Classes/UniSdkPlugin.swift
import Flutter
import UIKit

public class UniSdkPlugin: NSObject, FlutterPlugin, DCUniMPSDKEngineDelegate {
    var channel: FlutterMethodChannel?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "uni_sdk_plugin", binaryMessenger: registrar.messenger())
        let instance = UniSdkPlugin()
        instance.channel = channel
        registrar.addMethodCallDelegate(instance, channel: channel)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        
        print(call.method)
        if(call.method == "openUniMP"){
            let arguments = call.arguments as! Dictionary<String, Any>
            beforeOpenUniMP(appid: arguments["appId"] as! String, url: arguments["url"] as! String, version: arguments["version"] as! String, result: result)
        }else{
            result(FlutterMethodNotImplemented)
        }
    }
    
    public func onUniMPEventReceive(_ appid: String, event: String, data: Any, callback: @escaping DCUniMPKeepAliveCallback) {
        //监听小程序发送事件
        if(event == "getUserInfo"){
            getUserInfo { result in
                if let result = result as? Dictionary<String, Any> {
                    let uid = result["uid"]
                    let token = result["token"]
                    callback(["uid": uid, "token":token], false)
                }
            }
            
        }
    }
    
    private func beforeOpenUniMP(appid: String, url: String, version: String, result: @escaping FlutterResult){
        //打开小程序
        //判断小程序是否已存在
        var vnName = "";
        if let versionInfo = DCUniMPSDKEngine.getUniMPVersionInfo(withAppid: appid) as? [String: Any]{
            vnName = versionInfo["name"] as? String ?? ""
        }
        if(DCUniMPSDKEngine.isExistsUniMP(appid) && vnName == version){
            openUniMP(appid: appid, result: result)
        }else{
//            if(!isFileExistsInDocuments(filename: URL(string: url)!.lastPathComponent)){
                //下载文件
                downloadFileAndSaveToDocuments(url: URL(string: url)!, result: result){ [weak self] in
                    print("file download success")
                    self?.installAndOpenUniMP(appid: appid, url: url, result: result)
                }
//            }else{
//                installAndOpenUniMP(appid: appid, url: url, result: result)
//            }
        }
    }
    
    func openUniMP(appid: String, result: @escaping FlutterResult) {
        let config = self.getUniMPConfiguration()
        DCUniMPSDKEngine.openUniMP(appid, configuration: config) { instance, error in
            if instance != nil {
                print("小程序打开成功")
                //self.uniMPInstance = instance
                result(true)
            } else {
                print(error as Any)
                result(false)
            }
        }
    }
    
    func installAndOpenUniMP(appid: String, url: String, result: @escaping FlutterResult){
        print("install uni mp")
        let docsDir = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true)[0]
        let filePath = docsDir + "/" + URL(string: url)!.lastPathComponent
        print(filePath)
        
        do{try DCUniMPSDKEngine.installUniMPResource(withAppid: appid, resourceFilePath: filePath, password: nil)
            let fileManager = FileManager.default
            try fileManager.removeItem(at: URL(string: filePath)!)
            self.openUniMP(appid: appid, result: result)
        }catch{
            print("load unimp error")
        }
        
    }
    
    func getUniMPConfiguration()->DCUniMPConfiguration{
        let config = DCUniMPConfiguration.init()
        config.extraData = ["arguments": "hello"]
        config.enableBackground = true
        return config
    }
    
    //判断文件是否存在
//    func isFileExistsInDocuments(filename: String) -> Bool {
//        let fileManager = FileManager.default
//        let documentsURL = fileManager.urls(for: .documentDirectory, in: .userDomainMask)[0]
//        let fileURL = documentsURL.appendingPathComponent(filename)
//
//        return fileManager.fileExists(atPath: fileURL.path)
//    }
    
    func downloadFileAndSaveToDocuments(url: URL, result: @escaping FlutterResult, completion:@escaping ()->Void) {
        //下载文件
        let session = URLSession.shared
        
        let task = session.downloadTask(with: url) { (temporaryURL, response, error) in
            if let error = error {
                print("Error downloading file: \(error)")
                result(false)
                return
            }
            
            guard let temporaryURL = temporaryURL else {
                print("Temporary URL is nil.")
                result(false)
                return
            }
            
            let fileManager = FileManager.default
            let documentsURL = fileManager.urls(for: .documentDirectory, in: .userDomainMask)[0]
            
            let destinationURL = documentsURL.appendingPathComponent(url.lastPathComponent)
            
            do {
                try fileManager.moveItem(at: temporaryURL, to: destinationURL)
                print("File downloaded and saved to: \(destinationURL.path)")
                DispatchQueue.main.sync {
                    completion()
                }
            } catch {
                result(false)
                print("Error saving file: \(error)")
            }
        }
        
        task.resume()
    }
}

extension UniSdkPlugin {
    /// 从Flutter中获取token数据
    func getUserInfo(result: @escaping FlutterResult) {
        channel?.invokeMethod("getUserInfo", arguments: nil, result: result)
    }
}
